### ISSUE TITLE
puzzle about custom sampler

### ISSUE THREAD

**Issue by xilinniao123** [jessekim-ck/2020AIChallenge04#2]
Thanks for your sharing! good job. i have a puzzle about your custom sampler,
#1、in your code 'triarch.py'
`data_loader = DataLoader(
            TrainDataset("train", "train_meta.csv", transform=transform),
            batch_size=self.args.batch_size*self.args.num_instances,
            num_workers=0,
            sampler=TrainSampler("train", "train_meta.csv", self.args.num_instances),
            drop_last=True,
            shuffle=False
        )`
#2、you use your custom sampler,and in your TrainSampler
`for idx in indices:
            pid = self.pid_list[idx]
            t = np.random.choice(self.pid_to_img[pid], size=self.num_instances, replace=False)
            ret.extend(t)`
you iterater every person face no,and generate num_instances images of this person face No, 
#3、last  in  train.py  file,you set 
`   arg_parser.add_argument("--batch_size", type=int, default=8)
    arg_parser.add_argument("--num_instances", type=int, default=16)`
in which num_instances is greater than batch_size,
===>so in my opinon,in one batch all images belone to the same person id,and in this case,then in one  batch you can't do TripletHardLoss,because in one batch all samples from the sample class(person)?? is what i said is true?or i mistake something? looking forward your explanation.

**Comment by jessekim-ck** [comment]
Hello @hitxujian ! Thank you for your opinion.

First of all, `TrainSampler` samples `num_instances` images(indices) per person identity, and stacks them on a list, in `__iter__` method.

Then when creating `DataLoader` object, the batch size is set to `args.batch_size * args.num_instances`, so that the data loader fetches images of `args.batch_size` independent people's, and `args.num_instances` images per person in a batch.

So when `batch_size` is set to 8 and `num_instances` is set to 16, total 8*16 images are fetched per batch, consisting of 8 identities of 16 images each.

If I'm wrong in my code, please let me know!

**Comment by hitxujian** [comment]
i aggree with your explanation，so I think in code `ret = list()

        # Randomly selects `num_instances` images for each identity and includes them in the index
        for idx in indices:
            pid = self.pid_list[idx]
            t = np.random.choice(self.pid_to_img[pid], size=self.num_instances, replace=False)
            ret.extend(t)`
extend should change to append??
